doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,257 @@
+
+* File sources
+
+  | mz_timestamp | offset |
+  |--------------|--------|
+  | ...          | ...    |
+
+* Kafka sources
+
+  | mz_timestamp | partition | offset |
+  |--------------|-----------|--------|
+  | ...          | ...       | ...    |
+
+* Kinesis sources
+
+  | mz_timestamp | shard_id | sequence_number |
+  |--------------|----------|-----------------|
+  | ...          | ...      | ...             |
+
+* PostgreSQL sources
+
+  | mz_timestamp | lsn |
+  |--------------|-----|
+  | ...          | ... |
+
+* S3 sources
+
+  | mz_timestamp | key | offset |
+  |--------------|-----|--------|
+  | ...          | ... | ...    |
+
+
+If `EXPOSE RECLOCKING` is not specified, and the source type is reclockable,
+Materialize will generate a name of the form `<source>_reclocking`. If that name
+already exists, Materialize will try names of the form `<source>_reclocking1`,
+`<source>_reclocking2`, ... in sequence until it finds an available name. This
+matches the logic used for generatic the name of a primary index for a table.
+
+The reclocked relation will be a new catalog item of type `Reclocking`:
+
+```rust
+// catalog.rs
+/// An item representing a source reclocking relation.
+struct Reclocking {
+    /// The ID of the source to which this reclocking applies.
+    source_id: GlobalId,
+    /// The type of the reclocking relation.
+    ///
+    /// This must be derived the type of the `source_id` source. It is
+    /// duplicated here, rather than being derived from the source on the fly,
+    /// due to limitations in the design of the catalog API. (The type of the
+    /// item must be derivable from the item itself, i.e., without access to the
+    /// rest of catalog).
+    desc: RelationDesc,
+}
+```
+
+`Reclocking` items will necessarily have a `GlobalId` in order to be included in
+the catalog; however this `GlobalId` will not be meaningful to the storage
+layer. (TODO: should it be? Should we instead have `CreateSources` accept
+separate `GlobalId`s for the remap and reclocked collections, for the source
+types that support both? We'd need to ensure that `Insert, `AllowCompaction`,
+et al. are not supported for remap collections.)
+
+### System catalog changes
+
+A new system table `mz_reclockings` will be added with the following schema:
+
+| id | name | source_id |
+|----|------|-----------|
+| .. | ...  | ...       |
+
+This table relates the name of a reclocking collection to a source. There is no
+need to describe the database or schema of the reclocking collection; like
+indexes, they always belong to the same schema as the relation with which they
+are associated.
+
+The `mz_objects` view will include items from `mz_reclockings` with a type of
+`reclocking`.
+
+## Open questions
+
+* Is storage planning to model the remap and reclocked collections as two
+  separate collections internally?
+
+* Should we use a term other than "reclocking" when talking about this feature
+  publicly and in the SQL syntax? "Timestamping"?
```

Comment:
I think we should use "timestamps" or "timestamping" publicly (e.g. `EXPOSE RECLOCKING` --> `EXPOSE TIMESTAMPS` and `mz_reclockings` --> `mz_timestamps`). My sense is that this is more intuitive to the end user, and reclocking would require some extra effort in explanation on our part.

Reply:
Yes, "reclocking" definitely requires more education! There is an inherent tradeoff here between introducing a brand new term ("reclocking") and co-opting an existing term ("timestamps"). The new term has the downside that it requires user education, but the upside that users can't come in with incorrect preexisting notions. Co-opting timestamps has the upside that it is more familiar to users, but the downside that you might think it refers to e.g. exposing Kafka timestamps, which it does not. 

I think the new term I'm going to advocate for is _progress_. This is the information about the _progress_ that Materialize has made in ingesting your source. I think that is the right intuition, and doesn't come with any external baggage. (Perhaps @frankmcsherry has complaints about how this is not quite the same thing as Timely's progress tracking, idk!)

Reply:
`PROGRESS` sounds great to me!